Modernize package metadata with pyproject.toml

cuddandr opened the pull request:
Fix a deprecation warning regarding the legacy install method by updating the package metadata and build specification (e.g. dependencies) to use `pyproject.toml` with a `setup.py` for building a dynamic dependency list.

Basically `pyproject.toml` contains most of the info, and `setup.py` is used to dynamically inspect the user's environment to install the correct version of CuPy based on the active CUDA version.

Also updated some of the metadata, such as the version, authors, and maintainers.

Let me know or edit the `pyproject.toml` to further update the package metadata. Since this is messing with the installation method, it should be tested by at least one other person.

cuddandr committed:
Modernize package metadata with pyproject.toml

Fix a deprecation warning by updating the package metadata and build
specification (e.g. dependencies) to use pyproject.toml with a setup.py
for building a dynamic dependency list.

Also updated some of the metadata, such as the version, authors, and
maintainers.

mjkramer committed:
Remove pynvjitlink

Perlmutter's CUDA driver now supports 12.4 directly

mjkramer commented:
Works in my tests. I added/tested a commit to remove pynvjitlink since it's no longer needed.

cuddandr committed:
Remove requirements.txt as it is no longer necessary

Deleted requirements.txt as all the dependencies are specified in
setup.py and the installation scripts do not use requirements.txt

Updated workflows/test.yml to remove the check for requirements.txt
